icicle/include/icicle/program/symbol.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+namespace icicle {
+
+  /**
+   * @brief Enum to represent all the operation supported by the Symbol class.
+   *
+   */
+  enum OpCode {
+    OP_ADD = 0,
+    OP_MULT,
+    OP_SUB,
+    OP_INV,
+
+    NOF_OPERATIONS,
+
+    OP_INPUT,
+    OP_CONST,
+  };
+
+  /**
+   * @brief a node at the data-flow grapsh (DFG) that represent a single operation
+   *
+   * This class contains a data about an operation done by the Symbol class.
+   * It saves oll the data to describe the operation done and point to its operands.
+   * At any given time a Sybol contains an Operation instance that represent all the
+   * calculation history done so far to reach the current value.
+   */
+
+  template <typename S>
+  class Operation
+  {
+  public:
+    OpCode m_opcode;
+    std::shared_ptr<Operation<S>> m_operand1; // 1st operand if exist
```

Comment:
(1) does this mean we won't ever have ops with more than 2 operands?
(2) this template won't support cases like op1 is scalar and op2 is EC point. Is that intentional? not interesting?

Reply:
(1) for now that's more than enough. If we'll need more, can be easily changed to a vector. For now, the operations are very simple
(2) Also here, I see no reason to support. No primitive in the near future that I know of will need that.